src/common/src/system_param/mod.rs

```diff
@@ -95,45 +116,50 @@ macro_rules! impl_system_params_to_kv {
     };
 }
 
+macro_rules! impl_derive_missing_fields {
+    ($({ $field:ident, $type:ty, $default:expr },)*) => {
+        fn derive_missing_fields(params: &mut SystemParams) {
+            $(
+                if params.$field.is_none() && let Some(v) = OverrideFromParams::$field(params) {
+                    params.$field = Some(v);
+                }
+            )*
+        }
+    };
+}
+
 // Derive deserialization from kv pairs.
 macro_rules! impl_system_params_from_kv {
     ($({ $field:ident, $type:ty, $default:expr },)*) => {
-        /// For each field in `SystemParams`, one of these rules apply:
-        /// - Up-to-date: Guaranteed to be `Some`. If it is not present, may try to derive it from previous
-        ///   versions of this field.
-        /// - Deprecated: Guaranteed to be `None`.
-        /// - Unrecognized: Not allowed.
-        pub fn system_params_from_kv(kvs: Vec<(impl AsRef<[u8]>, impl AsRef<[u8]>)>) -> Result<SystemParams> {
+        /// Try to deserialize deprecated fields as well.
+        /// Return error if there are unrecognized fields
+        pub fn system_params_from_kv<K, V>(mut kvs: Vec<(K, V)>) -> Result<SystemParams>
+        where
+            K: AsRef<[u8]> + Debug,
+            V: AsRef<[u8]> + Debug,
+        {
             let mut ret = SystemParams::default();
-            let mut expected_keys: HashSet<_> = [
-                $(key_of!($field),)*
-            ]
-            .iter()
-            .cloned()
-            .collect();
-            for (k, v) in kvs {
+            kvs.retain(|(k,v)| {
                 let k = std::str::from_utf8(k.as_ref()).unwrap();
                 let v = std::str::from_utf8(v.as_ref()).unwrap();
                 match k {
```

Comment:
We may only let the macro generate this `match` statement and define the function normally, so it'll be more friendly to editing.

Reply:
The difficulty is that the macro to generate `match` needs to capture `k, v, ret`. In [this commit](https://github.com/risingwavelabs/risingwave/pull/8260/commits/f81000a01206110b94c17047562a3984c01e65ee), I modified `for_all_params` and `for_all_undeprecated_params` to accept custom arguments, but it makes the macro definition and usage more complex.

Reply:
You might be right. 🥵 Let's prefer simplicity.